File: code/python/02-structured-output/function_calling_vs_structured.py
"""Function Calling vs. Structured Output: side-by-side comparison.

Runs the same sentiment-extraction task through both API paths on 5 test texts
and prints a per-text table plus a summary comparing success rate, total tokens,
and average latency for each method.

See docs/01-foundations/03-structured-output.md
  — "Function Calling vs. Structured Output: The Real Difference"
"""

from __future__ import annotations

import json
import os
import time
from dataclasses import dataclass, field
from typing import Literal, Optional

from openai import OpenAI
from pydantic import BaseModel, Field, ValidationError

MODEL = "gpt-4o"


class SentimentResponse(BaseModel):
    """Output schema shared by both extraction paths."""

    sentiment: Literal["positive", "negative", "neutral"]
    confidence: float = Field(..., ge=0, le=1)
    key_phrases: list[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Function-calling definition (Path A)
# ---------------------------------------------------------------------------

FUNCTION_TOOL: dict = {
    "type": "function",
    "function": {
        "name": "classify_sentiment",
        "description": "Classify the sentiment of the provided text.",
        "parameters": {
            "type": "object",
            "properties": {
                "sentiment": {
                    "type": "string",
                    "enum": ["positive", "negative", "neutral"],
                    "description": "Overall sentiment.",
                },
                "confidence": {
                    "type": "number",
                    "minimum": 0,
                    "maximum": 1,
                    "description": "Confidence score between 0 and 1.",
                },
                "key_phrases": {
                    "type": "array",
                    "items": {"type": "string"},
                    "maxItems": 5,
                    "description": "Key phrases that influenced the classification.",
                },
            },
            "required": ["sentiment", "confidence"],
        },
    },
}

# ---------------------------------------------------------------------------
# JSON-schema response_format definition (Path B)
# ---------------------------------------------------------------------------

_SCHEMA = SentimentResponse.model_json_schema()
# Strict mode requires additionalProperties: false and all properties in required.
_SCHEMA.setdefault("additionalProperties", False)
_SCHEMA.setdefault("required", list(_SCHEMA.get("properties", {}).keys()))

# ---------------------------------------------------------------------------
# Result container
# ---------------------------------------------------------------------------


@dataclass
class Result:
    method: str
    text: str
    success: bool
    sentiment: Optional[str] = None
    confidence: Optional[float] = None
    prompt_tokens: int = 0
    completion_tokens: int = 0
    latency_ms: float = 0.0
    error: Optional[str] = None


# ---------------------------------------------------------------------------
# Extraction functions
# ---------------------------------------------------------------------------


def extract_function_calling(text: str, client: OpenAI) -> Result:
    """Extract sentiment via function calling (Path A)."""
    start = time.perf_counter()
    try:
        response = client.chat.completions.create(
            model=MODEL,
            messages=[{"role": "user", "content": text}],
            tools=[FUNCTION_TOOL],
            tool_choice={
                "type": "function",
                "function": {"name": "classify_sentiment"},
            },
        )
        latency_ms = (time.perf_counter() - start) * 1000
        tool_call = response.choices[0].message.tool_calls[0]
        args = json.loads(tool_call.function.arguments)
        parsed = SentimentResponse(**args)
        return Result(
            method="function_calling",
            text=text,
            success=True,
            sentiment=parsed.sentiment,
            confidence=parsed.confidence,
            prompt_tokens=response.usage.prompt_tokens,
            completion_tokens=response.usage.completion_tokens,
            latency_ms=latency_ms,
        )
    except Exception as exc:
        return Result(
            method="function_calling",
            text=text,
            success=False,
            error=str(exc),
            latency_ms=(time.perf_counter() - start) * 1000,
        )


def extract_structured_output(text: str, client: OpenAI) -> Result:
    """Extract sentiment via structured output / json_schema (Path B)."""
    start = time.perf_counter()
    try:
        response = client.chat.completions.create(
            model=MODEL,
            messages=[{"role": "user", "content": text}],
            response_format={
                "type": "json_schema",
                "json_schema": {
                    "name": "sentiment_response",
                    "schema": _SCHEMA,
                    "strict": True,
                },
            },
        )
        latency_ms = (time.perf_counter() - start) * 1000
        raw = response.choices[0].message.content or ""
        parsed = SentimentResponse.model_validate_json(raw)
        return Result(
            method="structured_output",
            text=text,
            success=True,
            sentiment=parsed.sentiment,
            confidence=parsed.confidence,
            prompt_tokens=response.usage.prompt_tokens,
            completion_tokens=response.usage.completion_tokens,
            latency_ms=latency_ms,
        )
    except Exception as exc:
        return Result(
            method="structured_output",
            text=text,
            success=False,
            error=str(exc),
            latency_ms=(time.perf_counter() - start) * 1000,
        )


# ---------------------------------------------------------------------------
# Test texts
# ---------------------------------------------------------------------------

TEST_TEXTS = [
    "I absolutely love this product! Best purchase I've ever made.",
    "This is absolutely terrible. Complete waste of money.",
    "It arrived. Haven't tried it yet.",
    "Oh sure, because *that's* exactly what I needed — another broken feature.",  # sarcastic
    "Ce produit est fantastique, je le recommande vivement.",  # French
]


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------


def main() -> None:
    client = OpenAI(api_key=os.environ["OPENAI_API_KEY"])

    fc_results: list[Result] = []
    so_results: list[Result] = []

    for text in TEST_TEXTS:
        fc = extract_function_calling(text, client)
        so = extract_structured_output(text, client)
        fc_results.append(fc)
        so_results.append(so)

        agree = (
            fc.sentiment == so.sentiment
            if fc.success and so.success
            else "N/A"
        )
        print(f"\nText: {text[:65]!r}")
        print(f"  {'Method':<22} {'Result':<12} {'Tokens':>8} {'Latency':>10}  Status")
        print(f"  {'-'*58}")
        print(
            f"  {'function_calling':<22} {fc.sentiment or 'FAIL':<12} "
            f"{fc.prompt_tokens + fc.completion_tokens:>8} "
            f"{fc.latency_ms:>9.0f}ms  {'✓' if fc.success else '✗'}"
        )
        print(
            f"  {'structured_output':<22} {so.sentiment or 'FAIL':<12} "
            f"{so.prompt_tokens + so.completion_tokens:>8} "
            f"{so.latency_ms:>9.0f}ms  {'✓' if so.success else '✗'}"
        )
        print(f"  Results agree: {agree}")

    # Summary
    print("\n" + "=" * 58)
    print("SUMMARY")
    print("=" * 58)
    for label, results in [
        ("Function Calling", fc_results),
        ("Structured Output", so_results),
    ]:
        successes = [r for r in results if r.success]
        total_tok = sum(r.prompt_tokens + r.completion_tokens for r in successes)
        avg_tok = total_tok // max(len(successes), 1)
        avg_lat = sum(r.latency_ms for r in results) / max(len(results), 1)
        print(f"\n{label}:")
        print(f"  Success rate : {len(successes)}/{len(results)}")
        print(f"  Avg tokens   : {avg_tok}")
        print(f"  Avg latency  : {avg_lat:.0f}ms")


if __name__ == "__main__":
    main()

File: code/go/02-structured-output/function_calling_vs_structured.go
// Function Calling vs. Structured Output: side-by-side comparison.
//
// Runs the same sentiment-extraction task through both API paths on 5 test
// texts and prints a per-text table plus a summary comparing success rate,
// total tokens, and average latency for each method.
//
// Port of code/python/02-structured-output/function_calling_vs_structured.py
// Run: go run . -mode=compare
//
// See docs/01-foundations/03-structured-output.md
// — "Function Calling vs. Structured Output: The Real Difference"
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strings"
	"time"

	openai "github.com/sashabaranov/go-openai"
)

const fcsModel = "gpt-4o"

// fcsSentimentTool is the function-calling tool definition (Path A).
var fcsSentimentTool = openai.Tool{
	Type: openai.ToolTypeFunction,
	Function: &openai.FunctionDefinition{
		Name:        "classify_sentiment",
		Description: "Classify the sentiment of the provided text.",
		Parameters: map[string]interface{}{
			"type": "object",
			"properties": map[string]interface{}{
				"sentiment": map[string]interface{}{
					"type":        "string",
					"enum":        []string{"positive", "negative", "neutral"},
					"description": "Overall sentiment.",
				},
				"confidence": map[string]interface{}{
					"type":        "number",
					"minimum":     0,
					"maximum":     1,
					"description": "Confidence score between 0 and 1.",
				},
				"key_phrases": map[string]interface{}{
					"type":        "array",
					"items":       map[string]interface{}{"type": "string"},
					"maxItems":    5,
					"description": "Key phrases that influenced the classification.",
				},
			},
			"required": []string{"sentiment", "confidence"},
		},
	},
}

// fcsSentimentSchema is the JSON Schema for structured output / json_schema (Path B).
// Strict mode requires additionalProperties: false and all properties listed in required.
var fcsSentimentSchema = map[string]interface{}{
	"type": "object",
	"properties": map[string]interface{}{
		"sentiment": map[string]interface{}{
			"type": "string",
			"enum": []string{"positive", "negative", "neutral"},
		},
		"confidence": map[string]interface{}{
			"type":    "number",
			"minimum": 0,
			"maximum": 1,
		},
		"key_phrases": map[string]interface{}{
			"type":  "array",
			"items": map[string]interface{}{"type": "string"},
		},
	},
	"required":             []string{"sentiment", "confidence", "key_phrases"},
	"additionalProperties": false,
}

// fcsResult holds the outcome of one extraction attempt.
type fcsResult struct {
	method           string
	text             string
	success          bool
	sentiment        string
	confidence       float64
	promptTokens     int
	completionTokens int
	latencyMS        float64
	errStr           string
}

// fcsTestTexts is the shared set of test inputs for both extraction paths.
var fcsTestTexts = []string{
	"I absolutely love this product! Best purchase I've ever made.",
	"This is absolutely terrible. Complete waste of money.",
	"It arrived. Haven't tried it yet.",
	"Oh sure, because *that's* exactly what I needed — another broken feature.",
	"Ce produit est fantastique, je le recommande vivement.",
}

// fcsFunctionCalling extracts sentiment via function calling (Path A).
func fcsFunctionCalling(ctx context.Context, text string, client *openai.Client) fcsResult {
	start := time.Now()
	res := fcsResult{method: "function_calling", text: text}

	resp, err := client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model:    fcsModel,
		Messages: []openai.ChatCompletionMessage{{Role: openai.ChatMessageRoleUser, Content: text}},
		Tools:    []openai.Tool{fcsSentimentTool},
		ToolChoice: openai.ToolChoice{
			Type:     openai.ToolTypeFunction,
			Function: openai.ToolFunction{Name: "classify_sentiment"},
		},
	})
	res.latencyMS = float64(time.Since(start).Milliseconds())

	if err != nil {
		res.errStr = err.Error()
		return res
	}
	if len(resp.Choices) == 0 || len(resp.Choices[0].Message.ToolCalls) == 0 {
		res.errStr = "no tool call in response"
		return res
	}

	var args struct {
		Sentiment  string  `json:"sentiment"`
		Confidence float64 `json:"confidence"`
	}
	if err := json.Unmarshal([]byte(resp.Choices[0].Message.ToolCalls[0].Function.Arguments), &args); err != nil {
		res.errStr = err.Error()
		return res
	}

	res.success = true
	res.sentiment = args.Sentiment
	res.confidence = args.Confidence
	res.promptTokens = resp.Usage.PromptTokens
	res.completionTokens = resp.Usage.CompletionTokens
	return res
}

// fcsStructuredOutput extracts sentiment via structured output / json_schema (Path B).
func fcsStructuredOutput(ctx context.Context, text string, client *openai.Client) fcsResult {
	start := time.Now()
	res := fcsResult{method: "structured_output", text: text}

	schemaBytes, _ := json.Marshal(fcsSentimentSchema)

	resp, err := client.CreateChatCompletion(ctx, openai.ChatCompletionRequest{
		Model:    fcsModel,
		Messages: []openai.ChatCompletionMessage{{Role: openai.ChatMessageRoleUser, Content: text}},
		ResponseFormat: &openai.ChatCompletionResponseFormat{
			Type: openai.ChatCompletionResponseFormatTypeJSONSchema,
			JSONSchema: &openai.ChatCompletionResponseFormatJSONSchema{
				Name:   "sentiment_response",
				Schema: json.RawMessage(schemaBytes),
				Strict: true,
			},
		},
	})
	res.latencyMS = float64(time.Since(start).Milliseconds())

	if err != nil {
		res.errStr = err.Error()
		return res
	}

	var parsed struct {
		Sentiment  string  `json:"sentiment"`
		Confidence float64 `json:"confidence"`
	}
	if err := json.Unmarshal([]byte(resp.Choices[0].Message.Content), &parsed); err != nil {
		res.errStr = err.Error()
		return res
	}

	res.success = true
	res.sentiment = parsed.Sentiment
	res.confidence = parsed.Confidence
	res.promptTokens = resp.Usage.PromptTokens
	res.completionTokens = resp.Usage.CompletionTokens
	return res
}

// RunFunctionCallingVsStructured compares function calling and structured output
// for sentiment extraction across five test texts.
func RunFunctionCallingVsStructured() {
	client := openai.NewClient(os.Getenv("OPENAI_API_KEY"))
	ctx := context.Background()

	var fcResults, soResults []fcsResult

	for _, text := range fcsTestTexts {
		fc := fcsFunctionCalling(ctx, text, client)
		so := fcsStructuredOutput(ctx, text, client)
		fcResults = append(fcResults, fc)
		soResults = append(soResults, so)

		agree := "N/A"
		if fc.success && so.success {
			if fc.sentiment == so.sentiment {
				agree = "true"
			} else {
				agree = "false"
			}
		}

		preview := text
		if len(preview) > 65 {
			preview = preview[:65]
		}
		fmt.Printf("\nText: %q\n", preview)
		fmt.Printf("  %-22s %-12s %8s %10s  Status\n", "Method", "Result", "Tokens", "Latency")
		fmt.Println("  " + strings.Repeat("-", 58))

		fcSentiment, soSentiment := fc.sentiment, so.sentiment
		if !fc.success {
			fcSentiment = "FAIL"
		}
		if !so.success {
			soSentiment = "FAIL"
		}
		fcStatus, soStatus := "✓", "✓"
		if !fc.success {
			fcStatus = "✗"
		}
		if !so.success {
			soStatus = "✗"
		}

		fcTok := fc.promptTokens + fc.completionTokens
		soTok := so.promptTokens + so.completionTokens
		fmt.Printf("  %-22s %-12s %8d %9.0fms  %s\n", "function_calling", fcSentiment, fcTok, fc.latencyMS, fcStatus)
		fmt.Printf("  %-22s %-12s %8d %9.0fms  %s\n", "structured_output", soSentiment, soTok, so.latencyMS, soStatus)
		fmt.Printf("  Results agree: %s\n", agree)
	}

	// Summary
	fmt.Printf("\n%s\n", strings.Repeat("=", 58))
	fmt.Println("SUMMARY")
	fmt.Println(strings.Repeat("=", 58))

	type resultPair struct {
		label   string
		results []fcsResult
	}
	for _, p := range []resultPair{
		{"Function Calling", fcResults},
		{"Structured Output", soResults},
	} {
		var successes []fcsResult
		totalTok := 0
		for _, r := range p.results {
			if r.success {
				successes = append(successes, r)
				totalTok += r.promptTokens + r.completionTokens
			}
		}
		avgTok := 0
		if len(successes) > 0 {
			avgTok = totalTok / len(successes)
		}
		var totalLat float64
		for _, r := range p.results {
			totalLat += r.latencyMS
		}
		avgLat := 0.0
		if len(p.results) > 0 {
			avgLat = totalLat / float64(len(p.results))
		}
		fmt.Printf("\n%s:\n", p.label)
		fmt.Printf("  Success rate : %d/%d\n", len(successes), len(p.results))
		fmt.Printf("  Avg tokens   : %d\n", avgTok)
		fmt.Printf("  Avg latency  : %.0fms\n", avgLat)
	}
}
